### PR TITLE
Add support for alternative loading strategies

### LINKED ISSUE #440: Support for alternative loading strategies
_**Apologies if this is documented somewhere, but I have not been able to find anything about it.**_

**Is your feature request related to a problem? Please describe.**

Not really a problem, but more of an inconvenience. It would be nice to be able to specify the [Loading technique](https://docs.sqlalchemy.org/en/13/orm/loading_relationships.html) without having to drop to using the query builder.

**Describe the solution you'd like**

I would like to be able to specify the loading technique at the time I choose to load relationships.

Looking at the docs on [Collections](https://mikro-orm.io/docs/collections) and seeing how relationships are currently loaded, one potential API for this could be something like:

```ts
type Strategy = 'lazy' | 'joined' | 'subquery' | 'selectin';
```

```ts
const author = orm.em.findOne(Author, '...', ['books', 'publishers']); // current

const author = orm.em.findOne(Author, '...', {
  books: {
    strategy: 'joined',
  },
  publishers: {
    strategy: 'subquery',
  },
}); // suggested

await author.books.init({
  strategy: 'joined',
});
```

Similar to SQLAlchemy, you could even specify the load type at the relationship level:

```ts
@Entity()
export class Book {
  @ManyToOne({
    entity: () => Author,
    strategy: 'joined',
  })
  author!: Author;
}
```

**Describe alternatives you've considered**

As mentioned above, the alternative right now seems to be to use the query builder.

**Additional context**

This feature seems fairly standard to ORMs. Some examples include [SQLAlchemy](https://docs.sqlalchemy.org/en/13/orm/loading_relationships.html) and [Objection.js](https://vincit.github.io/objection.js/api/query-builder/eager-methods.html#withgraphfetched)

---

Thanks for the great project btw!

### PR DESCRIPTION
**Pull request by willsoto**
This adds support for alternate load strategies. As of now, only one additional strategy is supported: `joined`. Currently, the only way to specify the strategy is on on the decorator for the relationship:

```ts
@Entity()
export class Author2 {
  @OneToMany({
    entity: () => Book2,
    mappedBy: 'author',
    orderBy: { title: QueryOrder.ASC },
    strategy: LoadStrategy.JOINED, // new property
  })
  books!: Collection<Book2>;
}
```
---

- [ ] Tests
- [ ] Docs
- [ ] Cleanup

Closes #440

### CODE REVIEW

File: packages/knex/src/AbstractSqlDriver.ts [hunk 1]
```diff
@@ -63,15 +63,60 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
...
+      .select(selects)
       .populate(options.populate)
       .where(where as Dictionary)
       .orderBy(options.orderBy!)
-      .limit(1)
+      // .limit(1)
       .setLockMode(options.lockMode)
       .withSchema(options.schema)
-      .execute('get');
+      .execute('all');
```
Comment (willsoto) [inline packages/knex/src/AbstractSqlDriver.ts:119]: When joining records (like in a one-to-many) you necessarily get back >1 row, so that is why these changes were made.

File: packages/core/src/EntityManager.ts [hunk 2]
```diff
@@ -155,13 +155,19 @@ export class EntityManager<D extends IDatabaseDriver = IDatabaseDriver> {
...
+    }, this.transactionContext);
 
     if (!data) {
       return null;
     }
 
-    entity = this.merge(entityName, data as EntityData<T>, options.refresh) as T;
+    const entityData = !Array.isArray(data) ? [data] : data;
+
+    entity = this.merge(entityName, entityData[0] as EntityData<T>, options.refresh) as T;
```
Comment (willsoto) [inline packages/core/src/EntityManager.ts:170]: This is the point at which the raw query results actually become instances of things, so I'd imagine this is where I should also map any additional properties that come through?

File: packages/core/src/EntityManager.ts [hunk 3]
```diff
@@ -155,13 +155,19 @@ export class EntityManager<D extends IDatabaseDriver = IDatabaseDriver> {
...
+      ...options,
+      populate: preparedPopulate,
+    }, this.transactionContext);
 
     if (!data) {
       return null;
     }
 
-    entity = this.merge(entityName, data as EntityData<T>, options.refresh) as T;
+    const entityData = !Array.isArray(data) ? [data] : data;
```
Comment (B4nan) [inline packages/core/src/EntityManager.ts:168]:
```suggestion
    const entityData = Utils.asArray(data);
```

File: packages/core/src/EntityManager.ts (hunk 2 above)
Comment (B4nan) [reply packages/core/src/EntityManager.ts:170]: as noted above, `em.merge()` should not serve this purpose, its method for merging entities = making them managed. 

`DatabaseDriver.mapResult()` is the right place I think, probably by overriding the implementation in `AbstractSqlDriver`. that is what gets called when you use QB.getResult() - in fact it uses `em.map()` which first uses driver's `mapResult` followed by `em.merge()`. Then we could probably rewrite the `AbstractSqlDriver.findOne` so it uses `QB.getResult()` (or actually `QB.getSingleResult()` for `findOne`).

or maybe it should happen in the `Hydrator` (that is used in `EntityFactory`).

File: packages/knex/src/AbstractSqlDriver.ts (hunk 1 above)
Comment (B4nan) [reply packages/knex/src/AbstractSqlDriver.ts:119]: although this makes sense, i would rather do the mapping via QB as suggested above, as I dont like the idea of driver returning multiple rows for findOne. currently driver returns POJOs that are converted to entities via the `em.merge()` call, but we could return entities directly I hope (at least for `AbstractSqlDriver` - we can keep the `em.merge()` call sitting on the EM level to maintain compatibility with mongo driver. 

or we can narrow it for all drivers, as we are aiming v4 with this, we can afford some breaking changes.

File: packages/knex/src/AbstractSqlDriver.ts [hunk 4]
```diff
@@ -63,15 +63,60 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
       options.fields.unshift(pk);
     }
 
-    return this.createQueryBuilder(entityName, ctx, !!ctx)
-      .select(options.fields || '*')
+    const joinedLoads: string[] = options.populate.filter((field) => {
+      const relation = meta.properties[field];
+
+      return relation.strategy === 'joined';
+    });
```
Comment (willsoto) [inline packages/knex/src/AbstractSqlDriver.ts:70]: Related to this, I am wondering where the best place would to convert `populate` from `string[]` to `PopulateOptions[]` so that it can be passed around to all subsequent call sites who might need it.

My thinking is that we will get `populate`, `find`, `findOne` etc for free if we can standardize the data structure early on.

```ts
// all the various signatures for populate
populate = "books";
populate = ["books", "tags"];
populate = [
  {
    field: "books",
    strategy: "joined",
  },
  "tags",
];
```

into

```ts
populate = [
  {
    field: "books",
    strategy: "joined",
  },
  {
    field: "tags",
    strategy: "selectin",
  },
];
```

File: packages/knex/src/AbstractSqlDriver.ts [hunk 5]
```diff
@@ -63,15 +63,60 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
...
+
+        Object.values(properties)
+          .filter(({ reference }) => reference === 'scalar')
+          .forEach((descriptor) => {
+            const [fieldName] = descriptor.fieldNames;
+            const aliasedProp = `"${alias}"."${fieldName}" as "${alias}_${fieldName}"`;
+
+            selects.push(aliasedProp);
+            qb.select(aliasedProp);
+            qb.join(relationName, alias);
```
Comment (willsoto) [inline packages/knex/src/AbstractSqlDriver.ts:103]: I am not sure if I was doing something wrong, but if I don't call `select` prior to calling `join`, I get an error from the `join` call since there is nothing in the `_aliasMap` yet.

Is this expected or am I doing something? I guess this isn't necessarily a problem, I'm mostly curious.

File: packages/knex/src/AbstractSqlDriver.ts (hunk 1 above)
Comment (willsoto) [reply packages/knex/src/AbstractSqlDriver.ts:119]: > i would rather do the mapping via QB as suggested above, as I dont like the idea of driver returning multiple rows for findOne.

To be clear, are you suggesting that we convert the POJOs into actually entities in the QB? Or convert the POJO into the expected structure. Eg we take the flat object we get (from op) and turn it into:

```js
const author = {
  createdAt: "2020-04-26T22:32:53.482Z",
  updatedAt: "2020-04-26T22:32:53.490Z",
  name: "Albert Camus",
  email: "albert.camus@email.com",
  age: null,
  termsAccepted: false,
  optional: null,
  identities: null,
  born: null,
  bornTime: null,
  id: 1,
  books: [
    {
      uuid_pk: "1ce80133-0379-440d-b92e-e18e93428434",
      created_at: "2020-04-26T22:32:53.482Z",
      title: "The Stranger",
      perex: null,
      price: null,
      double: null,
      meta: null,
    },
  ],
};
```

> but we could return entities directly I hope (at least for AbstractSqlDriver - we can keep the em.merge() call sitting on the EM level to maintain compatibility with mongo driver.

If we return the above data structure, the `AbstractSqlDriver` could then map over the new structure, instantiate entities and return those back to the EM.

Does that sound right?

File: packages/knex/src/AbstractSqlDriver.ts (hunk 5 above)
Comment (B4nan) [reply packages/knex/src/AbstractSqlDriver.ts:103]: Kinda expected, we could work that around by making the join evaluation lazy (in the finalize method), but I don't think its an actual issue for us.

File: packages/knex/src/AbstractSqlDriver.ts (hunk 4 above)
Comment (B4nan) [reply packages/knex/src/AbstractSqlDriver.ts:70]: Agreed that it should happen early as possible, probably right after the `canPopulate` call, or maybe even inside it (once we validate it is ok). So probably at the EM level, I guess the `preparePopulate` method can be a good fit here.

File: packages/knex/src/AbstractSqlDriver.ts (hunk 1 above)
Comment (B4nan) [reply packages/knex/src/AbstractSqlDriver.ts]: First of all, you can already see that your current payload is something in the middle, as the root entity properties are already mapped to the entity prop names (camel case), while the book entity props are intact (snake case as from the db). This is what happens in the `mapResult` method of the driver, and that is where I think we should handle the child entities too so we do not have this kind of mixed state (some props mapped, some not).

About the QB, it already supports mapping to entities, that is what the `getResult()` method is for, which calls `em.map()` (which calls `driver.mapResult()` followed by `em.merge()`). So we might leverage that probably, that was my suggestion. 

So to wrap up, I think the `driver.mapResult()` should be where we convert the output to a real graph (not entities, but POJOs), so for findOne we still have single object in the return value (btw `mapResult()` is also used when doing `qb.execute()`)

File: packages/core/src/EntityManager.ts [hunk 6]
```diff
@@ -583,13 +591,41 @@ export class EntityManager<D extends IDatabaseDriver = IDatabaseDriver> {
...
+    const meta = this.metadata.get(entityName);
+
+    if (Array.isArray(populate)) {
+      return populate.map((option) => {
+        if (typeof option === 'string') {
+          const relation = meta.properties[option];
+
+          return {
+            field: option,
+            // TODO: Ideally we specify the default strategy somewhere else (maybe when the decorator is defined?)
```
Comment (B4nan) [inline packages/core/src/EntityManager.ts:611]: yes, the defaults should be set during the discovery (not in decorator, as there are other ways to define entities than just decorators). 

this should ideally happen in `EntitySchema` helper, that is also used internally in all cases (with all metadata providers).

File: packages/core/src/drivers/IDatabaseDriver.ts [hunk 7]
```diff
@@ -89,6 +89,8 @@ export interface FindOneOptions {
   schema?: string;
 }
 
+export type Populate = (string | PopulateOptions)[] | boolean;
```
Comment (B4nan) [inline packages/core/src/drivers/IDatabaseDriver.ts:92]: there is actually one thing i forgot to mention, as part of v4 i would like to deal with the #214 somehow, which will also need different format for this. we will have to be able to have strong typing support there, so maybe something like an object or tuples. 

so the `PopulateOptions` should be fine hopefully for that usecase, but we will need to extend it to support nesting (currently it is possible to have `populate: ['author.books.tags']`)

this is just a heads up, i dont think you need to care much about it now

**Comment by willsoto** [comment]
@B4nan - Ok, we are getting closer now...I've gotten it to the point where the joined pojos are being transformed correctly as suggested via `mapResult`. By doing that, we get the conversion to entities for free (as far as I can tell) and there are no additional queries fired since those entities are loaded already.

File: packages/core/src/entity/enums.ts [hunk 8]
```diff
@@ -13,3 +13,8 @@ export enum Cascade {
   REMOVE = 'remove',
   ALL = 'all',
 }
+
+export enum LoadStrategy {
+  SELECTIN = 'selectin',
+  JOINED = 'joined'
+}
```
Comment (willsoto) [inline packages/core/src/entity/enums.ts:20]: I converted this to an `enum` for consistency since you use enums to represent so many other options on the decorators.

Comment (B4nan) [reply packages/core/src/entity/enums.ts:20]: good. what about renaming `SELECTIN = 'selectin'` to `SELECT_IN = 'select-in'`?

File: tests/JoinedLoads.test.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import { MikroORM } from '@mikro-orm/core';
+import { PostgreSqlDriver } from '@mikro-orm/postgresql';
+import { initORMPostgreSql, wipeDatabasePostgreSql } from './bootstrap';
+import { Author2, Book2 } from './entities-sql';
+
+describe('Joined loading', () => {
```
Comment (willsoto) [inline tests/JoinedLoads.test.ts:6]: I'd additionally like to test that the correct number of queries are fired, so any suggestions for that?

eg when joined loading one relation, we should see 1 query, not 2

File: packages/core/src/entity/enums.ts (hunk 8 above)
Comment (willsoto) [reply packages/core/src/entity/enums.ts:20]: no problem 👍

File: tests/JoinedLoads.test.ts (hunk 9 above)
Comment (B4nan) [reply tests/JoinedLoads.test.ts:6]: mock logger and assert the actual queries. i do it quite often, search for `mock.calls` in the tests

**Comment by willsoto** [comment]
Also, this code is really gross...I am basically inlining everything and plan to back through and refactor (I hope you understand as I am adding a feature while learning the code base 😅  )

**Comment by B4nan** [comment]
Yeah sure, no problem at all :] 

Btw I am rebasing the `dev` branch quite often with bug fixes from master, so you might need to create new branch for this and copy the changes via stash or something. But it can happen later, once we will be getting close to merge.

**Comment by willsoto** [comment]
I feel like the code is _mostly_ in a good place for a more in-depth review. I still have 17 failing tests locally and I am currently looking into why they are failing with my changes. Some of them don't seem related and others seem to be non-deterministic, but `dev` is currently passing so I am looking into them.

File: packages/knex/src/AbstractSqlDriver.ts [hunk 10]
```diff
@@ -187,18 +249,92 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
...
+    // alias all fields in the primary table
+    Object.values(meta.properties)
+      .filter(({ reference, persist }) => {
+        return reference === 'scalar' && persist !== false;
+      })
+      .forEach((descriptor) => {
+        const [fieldName] = descriptor.fieldNames;
+
+        selects.push(fieldName);
+        queryBuilder.select(fieldName);
```
Comment (B4nan) [inline packages/knex/src/AbstractSqlDriver.ts:317]: `qb.select()` does override, maybe you want `qb.addSelect()`? or do we even need that line?

File: packages/knex/src/AbstractSqlDriver.ts [hunk 11]
```diff
@@ -187,18 +249,92 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
...
+      });
+
+    joinedLoads.forEach((relationName) => {
+      const prop = meta.properties[relationName];
+      const entity = this.metadata.get(prop.type);
```
Comment (B4nan) [inline packages/knex/src/AbstractSqlDriver.ts:322]: i dont like the name `entity` here. usually i use `meta2` to point to the other side, not great, but better than calling it `entity` when its not entity but its metadata.

File: packages/knex/src/AbstractSqlDriver.ts [hunk 12]
```diff
@@ -187,18 +249,92 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
...
+      const { properties } = entity;
```
Comment (B4nan) [inline packages/knex/src/AbstractSqlDriver.ts:323]: and as this is the only place where you use the `entity` variable, i would suggest to inline it so we dont have to care about naming :] 

```typescript
const properties = this.metadata.get(prop.type).properties;
```

File: packages/knex/src/AbstractSqlDriver.ts [hunk 13]
```diff
@@ -187,18 +249,92 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
...
+
+      Object.values(properties)
+        .filter(({ reference, persist }) => reference === 'scalar' && persist !== false)
+        .forEach((descriptor) => {
+          const [fieldName] = descriptor.fieldNames;
```
Comment (B4nan) [inline packages/knex/src/AbstractSqlDriver.ts:328]: i see you really love destructing :D 

but here i dont see the point, this is too verbose, when all we want is to access the first array element.

File: packages/knex/src/AbstractSqlDriver.ts [hunk 14]
```diff
@@ -187,18 +249,92 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
...
+      });
+
+    joinedLoads.forEach((relationName) => {
+      const prop = meta.properties[relationName];
+      const entity = this.metadata.get(prop.type);
+      const { properties } = entity;
+
+      Object.values(properties)
+        .filter(({ reference, persist }) => reference === 'scalar' && persist !== false)
+        .forEach((descriptor) => {
```
Comment (B4nan) [inline packages/knex/src/AbstractSqlDriver.ts:327]: we use `prop` everywhere in the code base so please stick to that.
```suggestion
        .forEach(prop => {
```

File: packages/knex/src/query/QueryBuilder.ts [hunk 15]
```diff
@@ -85,6 +85,12 @@ export class QueryBuilder<T extends AnyEntity<T> = AnyEntity> {
     return this.init(QueryType.COUNT);
   }
 
+  selectAliasedField(field: string, schema: string, index?: number) {
```
Comment (B4nan) [inline packages/knex/src/query/QueryBuilder.ts:88]: this should probably live elsewhere (abstract sql driver i guess?), also the method does not select anything.

and it looks like you never use the `index` param - what should that be about?

File: packages/knex/src/query/QueryBuilder.ts [hunk 16]
```diff
@@ -45,7 +45,7 @@ export class QueryBuilder<T extends AnyEntity<T> = AnyEntity> {
               private readonly connectionType?: 'read' | 'write',
               private readonly em?: SqlEntityManager) { }
 
-  select(fields: string | KnexQueryBuilder | (string | KnexQueryBuilder)[], distinct = false): this {
+  select(fields: string | Field[], distinct = false): this {
```
Comment (B4nan) [inline packages/knex/src/query/QueryBuilder.ts:48]:
```suggestion
  select(fields: Field | Field[], distinct = false): this {
```

File: packages/knex/src/query/QueryBuilder.ts [hunk 17]
```diff
@@ -491,6 +498,12 @@ export class QueryBuilder<T extends AnyEntity<T> = AnyEntity> {
 
 }
 
+type KnexStringRef = Knex.Ref<string, {
```
Comment (B4nan) [inline packages/knex/src/query/QueryBuilder.ts:501]:
```suggestion
type KnexStringRef = Ref<string, {
```

(and import `Ref` from `knex` if its not yet there)

File: packages/knex/src/query/QueryBuilder.ts [hunk 18]
```diff
@@ -491,6 +498,12 @@ export class QueryBuilder<T extends AnyEntity<T> = AnyEntity> {
...
+  [alias: string]: string;
+}>;
+
+export type Field = (KnexStringRef | string | KnexQueryBuilder);
```
Comment (B4nan) [inline packages/knex/src/query/QueryBuilder.ts:505]: no need to wrap in parens, also i would move the `string` to the first place (looks weird between those knex types :P)

**Comment by B4nan** [comment]
Sorry for the delay, the code looks good, most of the comments are just code style suggestions so its aligned with the codebase. 

Btw I think we will need to explicitly list all the fields when selecting anyway, as I am planning to introduce lazy scalar properties in v4, so we will need to be able to blacklist something from the select clause instead of using `*`.

File: tests/JoinedLoads.test.ts [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+    Object.assign(orm.em.config, { logger });
+
+    (await orm.em.findOne(Author2, {
+      id: author2.id,
+    }, {
+      populate: ['books'],
+    }));
+
+    expect(mock.mock.calls.length).toBe(1);
+    expect(mock.mock.calls[0][0]).toMatch('select "e0"."id", "e0"."created_at", "e0"."updated_at", "e0"."name", "e0"."email", "e0"."age", "e0"."terms_accepted", "e0"."optional", "e0"."identities", "e0"."born", "e0"."born_time", "books"."uuid_pk" as "books_uuid_pk", "books"."created_at" as "books_created_at", "books"."title" as "books_title", "books"."perex" as "books_perex", "books"."price" as "books_price", "books"."double" as "books_double", "books"."meta" as "books_meta" from "author2" as "e0" inner join "book2" as "books" on "e0"."id" = "books"."author_id" where "e0"."id" = $1');
```
[the remark this answers is not in the document]
Comment (willsoto) [reply tests/JoinedLoads.test.ts:55]: Agreed. In one of my first passes I added a numeric suffix, removed it and then forgot to add it back.

I think `b1`, `c2`, `b3`, etc should be sufficient.

File: tests/QueryBuilder.test.ts [hunk 20]
```diff
@@ -457,7 +457,7 @@ describe('QueryBuilder', () => {
 
   test('select by m:n inverse side (that is not defined as property) via populate', async () => {
     const qb = orm.em.createQueryBuilder(Test2);
-    qb.select('*').populate(['publisher2_tests']).where({ 'publisher2_tests.Publisher2_owner': { $in: [ 1, 2 ] } }).orderBy({ 'publisher2_tests.id': QueryOrder.ASC });
+    qb.select('*').populate([{ field: 'publisher2_to_test2' }]).where({ 'publisher2_to_test2.Publisher2_owner': { $in: [ 1, 2 ] } }).orderBy({ 'publisher2_to_test2.id': QueryOrder.ASC });
```
[the remark this answers is not in the document]
Comment (willsoto) [reply tests/QueryBuilder.test.ts:460]: Ah yes, I remember having merge conflicts here and I didn't notice the actual name changing. 👍

File: packages/core/src/entity/EntityLoader.ts [hunk 21]
```diff
@@ -14,25 +15,25 @@ export class EntityLoader {
 
   constructor(private readonly em: EntityManager) { }
 
-  async populate<T extends AnyEntity<T>>(entityName: string, entities: T[], populate: string | string[] | boolean, where: FilterQuery<T> = {}, orderBy: QueryOrderMap = {}, refresh = false, validate = true, lookup = true): Promise<void> {
-    if (entities.length === 0 || populate === false) {
+  async populate<T extends AnyEntity<T>>(entityName: string, entities: T[], populate: PopulateOptions[], where: FilterQuery<T> = {}, orderBy: QueryOrderMap = {}, refresh = false, validate = true, lookup = true): Promise<void> {
```
[the remark this answers is not in the document]
Comment (willsoto) [reply packages/core/src/entity/EntityLoader.ts:18]: Unintentional on my part. But I am curious what `true` implies?

File: packages/knex/src/AbstractSqlDriver.ts (hunk 10 above)
Comment (willsoto) [reply packages/knex/src/AbstractSqlDriver.ts:317]: Let me check, `addSelect` might do what I want here..

File: packages/knex/src/AbstractSqlDriver.ts (hunk 11 above)
Comment (willsoto) [reply packages/knex/src/AbstractSqlDriver.ts:322]: `meta2` seems reasonable enough

File: packages/knex/src/AbstractSqlDriver.ts (hunk 13 above)
Comment (willsoto) [reply packages/knex/src/AbstractSqlDriver.ts:328]: Your preference here would be?

```
const fieldName = descriptor.fieldNames[0];
```

File: packages/knex/src/query/QueryBuilder.ts (hunk 15 above)
Comment (willsoto) [reply packages/knex/src/query/QueryBuilder.ts:88]: I actually think I was already thinking about [this](https://github.com/mikro-orm/mikro-orm/pull/517#discussion_r419172881) when I added that.

Any naming suggestions? `generateAliasForField`?

File: packages/core/src/entity/EntityLoader.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (B4nan) [reply packages/core/src/entity/EntityLoader.ts:18]: https://github.com/mikro-orm/mikro-orm/issues/160

File: packages/knex/src/AbstractSqlDriver.ts (hunk 13 above)
Comment (B4nan) [reply packages/knex/src/AbstractSqlDriver.ts:328]: my preference is to not introduce new variable when not needed (especially for vars that are used only once), i would just use the `prop.fieldNames[0]` where needed. 

sometimes it makes sense to improve readability (like when you compose complex conditions), here i dont think so.

File: packages/knex/src/query/QueryBuilder.ts (hunk 15 above)
Comment (B4nan) [reply packages/knex/src/query/QueryBuilder.ts:88]: yeah sounds good. i would be even less verbose (e.g. `getAliasedField()`), but it might not be that readable, idk, naming is hard :]

File: packages/core/src/EntityManager.ts [hunk 22]
```diff
@@ -579,13 +593,47 @@ export class EntityManager<D extends IDatabaseDriver = IDatabaseDriver> {
...
+          // By returning a field/relation that doesn't exist, validation code will handle that error upstream.
+          if (!relation) {
+            return {
+              field: option,
+            };
+          }
+
+          return {
+            field: option,
+            strategy: relation.strategy,
```
[the remark this answers is not in the document]
Comment (willsoto) [reply packages/core/src/EntityManager.ts:621]: Not totally clear on the question here 😬

File: packages/knex/src/query/QueryBuilder.ts (hunk 15 above)
Comment (willsoto) [reply packages/knex/src/query/QueryBuilder.ts:88]: So looking at this code more, the reason why I put it in the `QueryBuilder` is because it needs access to the `knex` instance and that is currently not public. 

I do agree with the rename, so I did that. Also, the `index` param is used:

```
const as = index ? `${schema}${index}_${field}` : `${schema}_${field}`;
```

File: packages/knex/src/AbstractSqlDriver.ts (hunk 10 above)
Comment (willsoto) [reply packages/knex/src/AbstractSqlDriver.ts]: `addSelect` did the trick here 👍

File: packages/knex/src/AbstractSqlDriver.ts [hunk 23]
```diff
@@ -187,18 +249,92 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
...
 
+  protected joinedLoads(meta: EntityMetadata, populate: PopulateOptions[]): string[] {
+    return populate.filter(({ field }) => {
+      const relation = meta.properties[field];
+
+      if (!relation) {
+        return false;
+      }
+
+      return relation.strategy === LoadStrategy.JOINED;
```
[the remark this answers is not in the document]
Comment (willsoto) [reply packages/knex/src/AbstractSqlDriver.ts:289]: Not clear on what you mean here 🙃

File: packages/knex/src/query/QueryBuilder.ts [hunk 24]
```diff
@@ -321,6 +322,10 @@ export class QueryBuilder<T extends AnyEntity<T> = AnyEntity> {
     return qb;
   }
 
+  getRefForField(field: string, schema: string, as: string) {
```
Comment (B4nan) [inline packages/knex/src/query/QueryBuilder.ts:325]: i can't find the thread, but this can definitely live in abstract sql driver as i suggested. you have the knex instance there accessible from `this.connection.getKnex()`. its stateless/singleton, same instance as in `QB.knex`.

Comment (B4nan) [reply packages/knex/src/query/QueryBuilder.ts:325]: also please rename `as` to `alias`

File: packages/knex/src/AbstractSqlDriver.ts [hunk 25]
```diff
@@ -69,19 +74,66 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
...
+    joinedLoads.forEach((relationName, index) => {
+      const prop = meta.properties[relationName];
+      const properties = this.metadata.get(prop.type).properties;
 
-    return this.rethrow(qb.execute('get'));
+      ret[relationName] = ret[relationName] || [];
+      const relationPojo = {};
+
+      Object.values(properties)
+        .filter(({ reference }) => reference === 'scalar')
```
Comment (B4nan) [inline packages/knex/src/AbstractSqlDriver.ts:123]: use enum

File: tests/JoinedLoads.test.ts [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+    Object.assign(orm.em.config, { logger });
+
+    await orm.em.findOneOrFail(Author2, {
+      id: author2.id,
+    }, {
+      populate: ['books'],
+    });
+
+    expect(mock.mock.calls.length).toBe(1);
+    expect(mock.mock.calls[0][0]).toMatch('select "e0"."id", "e0"."created_at", "e0"."updated_at", "e0"."name", "e0"."email", "e0"."age", "e0"."terms_accepted", "e0"."optional", "e0"."identities", "e0"."born", "e0"."born_time", "books"."uuid_pk" as "b0_uuid_pk", "books"."created_at" as "b0_created_at", "books"."title" as "b0_title", "books"."perex" as "b0_perex", "books"."price" as "b0_price", "books"."double" as "b0_double", "books"."meta" as "b0_meta" from "author2" as "e0" inner join "book2" as "books" on "e0"."id" = "books"."author_id" where "e0"."id" = $1');
```
Comment (B4nan) [inline tests/JoinedLoads.test.ts:57]: this still looks a bit wrong, you are aliasing `book2` table as `books`, which can still conflict if you populate such property twice (e.g. on different entities, but its still called `books`).

it should be aligned with the prefix, so `b0`.

personally i would even use the same counter, so we have `a0` and `b1` instead of `a0` and `b0`. 
1. i would expect that to be easier to maintain/code
2. i feel like it helps understanding what is the root entity there.

i would also change the root alias from static `e0` to be dependent on actual entity name (so really `a0` here for author entity), but we can do that later (i can imagine that change to require lots of test changes...), will be happy to do it myself

**Comment by B4nan** [comment]
Later today I will be merging support for computed fields via `@Formula()` decorator to `dev`, I believe there will be some changes necessary to support this too. It's basically about being able to define custom sql snippets that will be automatically added to the select clause:

https://blog.eyallupu.com/2009/07/hibernate-derived-properties.html

Would be great if you could then rebase this and make sure it works with joined strategy too. Will ping you here and point you to what probably needs to be addressed.

File: packages/knex/src/AbstractSqlDriver.ts [hunk 27]
```diff
@@ -32,15 +32,17 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
   async find<T extends AnyEntity<T>>(entityName: string, where: FilterQuery<T>, options?: FindOptions, ctx?: Transaction<KnexTransaction>): Promise<T[]> {
     const meta = this.metadata.get(entityName);
     options = { populate: [], orderBy: {}, ...(options || {}) };
-    options.populate = this.autoJoinOneToOneOwner(meta, options.populate as string[]);
+
+    const populate = options.populate as PopulateOptions[];
+    options.populate = this.autoJoinOneToOneOwner(meta, populate);
```
[the remark this answers is not in the document]
Comment (willsoto) [reply packages/knex/src/AbstractSqlDriver.ts:37]: `populate` is used on line 45 as well

[the remark this answers is not in the document]
Comment (willsoto) [reply packages/knex/src/AbstractSqlDriver.ts:37]: I need to type assert in additional places since `options.populate` is a much wider type

File: packages/knex/src/AbstractSqlDriver.ts [hunk 28]
```diff
@@ -208,18 +263,87 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
...
     return [...populate, ...toPopulate];
   }
 
+  protected joinedLoads(meta: EntityMetadata, populate: PopulateOptions[]): string[] {
+    return populate
+      .filter(({ field }) => meta.properties[field]?.strategy === LoadStrategy.JOINED)
+      .map(({ field }) => field);
+  }
+
+  protected processJoinedLoads<T extends AnyEntity<T>>(results: object[], joinedLoads: string[]): T {
```
Comment (B4nan) [inline packages/knex/src/AbstractSqlDriver.ts:288]: looking at the failed tests, and one of your problems is that here you always return object, even when no entity is found

so we should first check for empty array and return `null` in such case

**Comment by B4nan** [comment]
Support for formulas is merged: https://github.com/mikro-orm/mikro-orm/pull/553

Currently it works only with the root entity, as that is the only one that gets actually mapped. With your changes, we will need to load those formulas for joined (populated) entities too.

here is the part that does the job, implemented at the QB level:

https://github.com/mikro-orm/mikro-orm/pull/553/files#diff-e32c03fc87a03e74dd84d662768fbd4eR461-R469

To support this with joined strategy, we will need to adjust the aliased name (e.g. `price_taxed -> b1_price_taxed`) and we will need to make sure we add it to selects also for joined entities.

Should be quite easy, I can help with that once we have all tests passing.

**Comment by willsoto** [comment]
So I am down to 7 test failures. I have some ordering issues and then I have a feeling I am not handling `populate: true` correctly somewhere, but I can't see it. I will continue to work on this but let me know if you see anything...

**Comment by B4nan** [comment]
You still have wrongly rebased changes there, mainly the schema dumps are wrong, as well as pivot table names in the entities, you are missing some tables there and some have bad names. Looks like you just omitted many of the changes I made, but kept some parts of it. For me ~200 tests are failing. 

This will be really hard to fix, as you made changes to fixtures so you are actually getting false positives, your tests might be passing but the changes I made are not present in the code. It would be better if you checkout fresh brach from `dev` and use git stash to move the changes from this brach, so you clearly see all things that changed. 

If you are lost there I can try to do this myself (via new branch/PR, as fixing this would take me days probably, as its basically about checking all the changes I made in dev branch).

**Comment by willsoto** [comment]
I’ll go ahead and just close this PR and cherry pick my changes into a new branch. Are you going to continue rebasing dev?

**Comment by willsoto** [comment]
Closing in favor of #556

**Comment by B4nan** [comment]
Thanks, will look into this when I have some spare time. Will try to not rebase dev till this one is merged (or the new one actually).
